### PR TITLE
Improvement of parallel work

### PR DESCRIPTION
**Pull request by Alex2251**
Improvement of duplicates in parallel work. Previous realization did duplicates in parallel work.

*Issue #, if available:*
In parallel work function duplicates output if you use DataLoader with num_workers > 1

*Description of changes:*
Added functionality to check on parallel work. If solution works in parallel than we take number of workers and each worker returns unique sequence.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

**Please tag this pr with at least one of these labels to make our release process faster:** BREAKING, new feature, bug fix, other change, dev setup

### CODE REVIEW

**Comment by lostella** [comment]
Hi @Alex2251, thanks for this! By the looks of it, this might be related to #2594?

To fix the style checks failures:

```
pip install -U black
black src test
```

and commit the changes

**Comment by Alex2251** [comment]
Hi @lostella! Thanks for comment and this commit will solve #2594.

**Comment by lostella** [comment]
@Alex2251 while this fixes #2594, each worker will still process each entry it needs to skip, and will slow down because of it. The downstream network will not be invoked (and there may be some time saving because of it), but any lazy transformation will still be applied to each element of the dataset. I think this was also pointed out by @gorold in [this comment](https://github.com/awslabs/gluonts/issues/2594#issuecomment-1418637840)

**Comment by Alex2251** [comment]
@lostella Yes it is. I tried to explaine my opinion in #2594, I think it is critical bug because you have a lot of the same batches.

**Comment by jaheba** [comment]
I wonder if we should use that ``DataLoader`` at all.

As I understand it, it does three things:

* collect inputs into batches
* stack those batches
* turns things into ``torch.tensor``

That's something we can easily do ourselves and in fact already do elsewhere.

**Comment by Alex2251** [comment]
Additionaly this DataLoader prevent duplicates, today there are duplicates. It shuld be prevented to use such dataloader because many people and me first time thought that it works from a box in a right way in parallel mode.

**Comment by lostella** [comment]
I think @jaheba means torch’s `DataLoader`: whatever we use, as long as the data is accessible only as an iterable stream, if multiple processes are to be used then there must be some logic to handle it. I’m not sure that avoiding torch’s `DataLoader` buys us anything.

The solution in this PR solves an issue, but it’s not ideal. The ideal solution, as pointed out, is either to rely on a “map” (“indexable”) dataset, or at least on a dataset type that can be sharded so that each process iterates its own shard: we used to have something like the latter, but it was arguably a bit clunky.

**Comment by lostella** [comment]
@jaheba what torch’s `DataLoader` buys us, I think, is to be able to use PyTorch Lightning: whether one likes it or not, I think it’s better than having to maintain our own training loop.

The best code is no code!

**Comment by Alex2251** [comment]
I agree.

**Comment by jaheba** [comment]
> @jaheba what torch’s `DataLoader` buys us, I think, is to be able to use PyTorch Lightning: whether one likes it or not, I think it’s better than having to maintain our own training loop.
> 
> The best code is no code!

That's just not true.

**Comment by gorold** [comment]
torch's ```DataLoader``` helps handle the multiprocessing for datasets! It's really important for larger batch sizes and can speed up training by 10x (assuming the underlying dataset is sharded properly). Implementing custom dataloading tooling which supports multiprocessing and interops with pytorch lightning (and all the accelerators, e.g. deepspeed, ddp, etc.) may be quite complicated.

**Comment by jaheba** [comment]
We have had bad experiences when using `multiprocessing` for loading data in the past and have therefore removed it. Our current strategy is to make the data-loading pipeline as fast as possible, and looking at parallelising it in the future.

Something we should do is to have some simple benchmarks to look at pipeline throughput, so that we actually have some numbers that we can talk about.

**Comment by jaheba** [comment]
We've removed the `DataLoader` in #2628.

I am going to create an issue to track the re-introduction of parallelised data loaders.
